File: src/main/scala/com/yogpc/qp/container/ContainerQuarryLevel.java
package com.yogpc.qp.container;

import com.yogpc.qp.packet.PacketHandler;
import com.yogpc.qp.packet.quarry.LevelMessage;
import com.yogpc.qp.tile.TileQuarry;
import com.yogpc.qp.version.VersionUtil;
import net.minecraft.entity.player.EntityPlayer;
import net.minecraft.entity.player.EntityPlayerMP;
import net.minecraft.inventory.Container;
import net.minecraft.inventory.Slot;
import net.minecraft.item.ItemStack;

public class ContainerQuarryLevel extends Container {
    private final TileQuarry tile;

    public ContainerQuarryLevel(TileQuarry tile, EntityPlayer player) {
        this.tile = tile;
        int oneBox = 18;
        for (int h = 0; h < 3; h++) {
            for (int v = 0; v < 9; v++) {
                this.addSlotToContainer(new Slot(player.inventory, v + h * 9 + 9, 8 + v * oneBox, 84 + h * oneBox));
            }
        }
        for (int vertical = 0; vertical < 9; vertical++) {
            this.addSlotToContainer(new Slot(player.inventory, vertical, 8 + vertical * oneBox, 142));
        }
        if (!tile.getWorld().isRemote) {
            PacketHandler.sendToClient(LevelMessage.create(tile), (EntityPlayerMP) player);
        }
    }

    @Override
    public boolean canInteractWith(EntityPlayer playerIn) {
        return tile.getWorld().getTileEntity(tile.getPos()) == tile;
    }

    @Override
    public ItemStack transferStackInSlot(EntityPlayer playerIn, int index) {
        return VersionUtil.empty();
    }
}

File: src/main/scala/com/yogpc/qp/container/ContainerQuarryLevel.scala
package com.yogpc.qp.container

import com.yogpc.qp.packet.{IMessage, PacketHandler}
import com.yogpc.qp.tile.HasInv
import com.yogpc.qp.version.VersionUtil
import net.minecraft.entity.player.{EntityPlayer, EntityPlayerMP}
import net.minecraft.inventory.{Container, Slot}
import net.minecraft.tileentity.TileEntity

class ContainerQuarryLevel[T <: TileEntity with HasInv](tile: T, player: EntityPlayer)(implicit messageFunc: T => _ <: IMessage) extends Container {
  val oneBox = 18

  for (h <- 0 until 3; v <- 0 until 9) {
    this.addSlotToContainer(new Slot(player.inventory, v + h * 9 + 9, 8 + v * oneBox, 84 + h * oneBox))
  }

  for (vertical <- 0 until 9) {
    this.addSlotToContainer(new Slot(player.inventory, vertical, 8 + vertical * oneBox, 142))
  }

  if (!tile.getWorld.isRemote)
    PacketHandler.sendToClient(messageFunc(tile), player.asInstanceOf[EntityPlayerMP])

  override def canInteractWith(playerIn: EntityPlayer) = tile.getWorld.getTileEntity(tile.getPos) eq tile

  override def transferStackInSlot(playerIn: EntityPlayer, index: Int) = VersionUtil.empty
}
